xarray/backends/zarr.py

```
@@ -0,0 +1,554 @@
+                                                          name=name)
+    var = conventions.maybe_encode_nonstring_dtype(var, name=name)
+    var = conventions.maybe_default_fill_value(var)
+    var = conventions.maybe_encode_bools(var)
+    var = conventions.ensure_dtype_not_object(var, name=name)
+    var = conventions.maybe_encode_string_dtype(var, name=name)
+    return var
+
+
+class ZarrStore(AbstractWritableDataStore, DataStorePickleMixin):
+    """Store for reading and writing data via zarr
+    """
+
+    def __init__(self, store=None, mode='a', synchronizer=None, group=None,
+                 writer=None, autoclose=None):
+        self._mode = mode
+        self._synchronizer = synchronizer
+        self._group = group
+
+        # zarr stores don't need to be opened, closed, or synced.
+        # So what do we do with all this logical about openers?
+        if autoclose:
+            raise NotImplementedError('autoclose=True is not implemented '
+                                      'for the zarr backend')
+        self._autoclose = False
+        self._isopen = True
+
+        opener = functools.partial(_open_zarr_group, store=store,
+                                   synchronizer=synchronizer, group=group)
+        self._opener = opener
+        self.ds = self._opener(mode=mode)
+
+        # initialize hidden dimension attribute
+        if _DIMENSION_KEY not in self.ds.attrs:
+            self.ds.attrs[_DIMENSION_KEY] = {}
+
+        # do we need to define attributes for all of the opener keyword args?
+        super(ZarrStore, self).__init__(writer)
+
+    def open_store_variable(self, name, zarr_array):
+        # I don't see why it is necessary to wrap self.ds[name]
+        # zarr seems to implement the required ndarray interface
+        # TODO: possibly wrap zarr array in dask with aligned chunks
+        data = indexing.LazilyIndexedArray(ZarrArrayWrapper(name, self))
+        dimensions, attributes = _get_zarr_dims_and_attrs(zarr_array,
```

This should raise a graceful error message when the dimension key is not found.